### DISCUSSION TITLE
分享一个无损的最极速的将flv转化为mp4的操作方案[ffmpeg]

### DISCUSSION THREAD

**Discussion by Lichtmann** [the1812/Bilibili-Evolved#1472] [分享 / Share]
一直在B站上用IDM下载flv格式的视频,因为画面和声音都在一起,只要下一次就好. 
* 下载好的flv当然是转化成mp4才好.
* 直接改后缀这种憨憨行为是没有用的
* 用格式工厂转换就是脱裤子放屁的同时加上倒立.

网上搜了一波发现flv,Ts,mp4就是封装不一样.
* 借助ffmpeg基本上2s-4s就可以将90秒长度的4K-FLV无损转化.
* 环境配置: 电脑本地需要安装或者解压ffmpeg的库以便调用.

之前我写好bat后,向每个下载目录里丢一个bat的快捷方式,然后可以将flv直接拖到快捷方式上转化.
*  昨天发现通过[Open with++](https://github.com/stax76/OpenWithPlusPlus)这个小工具可以实现右键FLV文件调出菜单栏执行Bat,很方便.
* 然后写了个目录循环遍历,点击一次就可以将同目录下所有FLV按顺序转化完成.
* 十分高效 快捷

# 核心代码
就是告诉ffmpeg复制视频流和音频流并切换为mp4封装:
ffmpeg -i "输入文件名" -vcodec copy -acodec copy "输出文件名"

## bat

```bat
@ echo off
::Set不能有空格
Set ffm=C:\SoftwareVideo\ffmpeg\bin\ffmpeg.exe 

for /r %~dp1 %%f in (*.flv) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_BilibiliFLV.mp4"
	if	"%2" == "replace" (
		del %%f
	)
)
for /r %~dp1 %%f in (*.ts) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_TS.mp4"
	if	"%2" == "replace" (
		del %%f
	)
)
pause
end
```

![Screenshot 2021-01-15 023425](https://user-images.githubusercontent.com/29507184/104669391-3828f680-56da-11eb-9903-801159784cf2.png)

**Comment by wangwang-code** [comment]
给这位大佬点👍
我对bat改了下

```
@ echo off
::Set不能有空格
Set ffm=C:\Users\Administrator\Desktop\ffmpeg.exe

for /r %~dp1 %%f in (*.flv) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_BilibiliFLV.mp4"
)
for /r %~dp1 %%f in (*.ts) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_TS.mp4"
)
exit
end
```
这个脚本不会删除源文件，执行完自动退出
另外，Open with++如果想让右键直接显示并调用bat
就只勾选Show for directories选项 ↓
![QQ截图20210411230029](https://user-images.githubusercontent.com/64469117/114309492-dbdb4480-9b19-11eb-99fd-76d050c1ebe7.png)
这样当你右键时
![QQ截图20210411230235](https://user-images.githubusercontent.com/64469117/114309523-03321180-9b1a-11eb-9a5a-ab8c6b99e014.png)
这就很方便了，不是吗

至于两三秒转化，得看你的硬盘
反正在固态里转化
啪，就闪了下，转好了
重新封装基本对CPU没要求（别打脸啊）

**Reply by Lichtmann** [reply]
你学会了👍
其他都是个人使用习惯的区别
我右键直接显示并调用的是单体操作
把批处理放到子菜单中了
Copy就是转化后保留源文件, Replace就是删除源文件
加一个Pause就是想看看整个过程有没有啥异常 哈哈哈

话说,舞蹈区的小姐姐真好看!

**Reply by wangwang-code** [reply]
没什么，就是看到有del就想到跟删除文件有关

不过有个想法，命令行在后台运行，自动输出log到目录下（防止幺蛾子），这样适合有大量FLV转化，但又不想影响做其他事情
右键点下就不用管的

**Reply by wangwang-code** [reply]
奋战1小时后终于完成了

> 不过有个想法，控制台在后台运行，自动输出日志到目录下（防止幺蛾子），这样适合有大量FLV转换，但又不想影响做其他事情

**Comment by wangwang-code** [comment]
叮~
后台运行+不删除文件+日志
```
@ echo off
if "%1"=="hide" goto CmdBegin
start mshta vbscript:createobject("wscript.shell").run("""%~0"" hide",0)(window.close)&&exit
:CmdBegin
::Set不能有空格
Set ffm=C:\Users\Administrator\Desktop\ffmpeg.exe

for /r %~dp1 %%f in (*.flv) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_BilibiliFLV.mp4" 2>> "BilibiliFLV.log"
)
for /r %~dp1 %%f in (*.ts) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_TS.mp4" 2>> "TS.log"
)
exit
end
```
原理就是cmd重定向
ffmpeg 的日志输出全部都是用标准错误输出，即stderr。
所以实际只有“2>”起到效果。
这里改成“2>>”就不会覆盖文件
如果想 ~后台运行~+删除文件+日志
```
@ echo off
::Set不能有空格
Set ffm=C:\Users\Administrator\Desktop\ffmpeg.exe

for /r %~dp1 %%f in (*.flv) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_BilibiliFLV.mp4" 2>> "BilibiliFLV.log"
	if	"%2" == "replace" (
		del %%f
	)
)
for /r %~dp1 %%f in (*.ts) do (
	%ffm% -i "%%f" -vcodec copy -acodec copy "%%~dpnf_TS.mp4" 2>> "TS.log"
	if	"%2" == "replace" (
		del %%f
	)
)
exit
end
```
很遗憾，加上这段用来后台运行的，不能del掉文件
```
if "%1"=="hide" goto CmdBegin
start mshta vbscript:createobject("wscript.shell").run("""%~0"" hide",0)(window.close)&&exit
:CmdBegin
```
_给后人挖个坑，你们自己填_
